src/backend/distributed/commands/create_citus_local_table.c

```diff
@@ -0,0 +1,635 @@
+ * single placement is only allowed to be on the coordinator.
+ */
+static void
+CreateCitusLocalTable(Oid relationId)
+{
+	/* these checks should be done before acquiring the lock on the table */
+	EnsureCoordinator();
+	EnsureTableOwner(relationId);
+
+	/*
+	 * Lock target relation with an exclusive lock - there's no way to make
+	 * sense of this table until we've committed, and we don't want multiple
+	 * backends manipulating this relation.
+	 */
+	Relation relation = relation_open(relationId, ExclusiveLock);
+
+	ErrorIfUnsupportedCreateCitusLocalTable(relationId);
+
+	ObjectAddress tableAddress = { 0 };
+	ObjectAddressSet(tableAddress, RelationRelationId, relationId);
+
+	/*
+	 * Ensure dependencies first as we will create shell table on the other nodes
+	 * in the MX case.
+	 */
+	EnsureDependenciesExistOnAllNodes(&tableAddress);
+
+	/*
+	 * Make sure that existing reference tables have been replicated to all
+	 * the nodes such that we can create foreign keys and joins work
+	 * immediately after creation.
+	 */
+	EnsureReferenceTablesExistOnAllNodes();
+
+	List *shellTableDDLEvents = GetShellTableDDLEventsForCitusLocalTable(relationId);
+
+	char *relationName = get_rel_name(relationId);
+	Oid relationSchemaId = get_rel_namespace(relationId);
+
+	/* below we convert relation with relationId to the shard relation */
+	uint64 shardId = ConvertLocalTableToShard(relationId);
+
+	/*
+	 * As we retrieved the DDL commands necessary to create the shell table
+	 * from scratch, below we simply recreate the shell table executing them
+	 * via process utility.
+	 */
+	CreateShellTableForCitusLocalTable(shellTableDDLEvents);
+
+	/*
+	 * Get new relationId as the relation with relationId now points
+	 * to the shard relation.
+	 */
+	Oid shellRelationId = get_relname_relid(relationName, relationSchemaId);
+
+	/* assert that we created the shell table properly in the same schema */
+	Assert(OidIsValid(shellRelationId));
+
+	InsertMetadataForCitusLocalTable(shellRelationId, shardId);
+
+	/* foreign tables do not support TRUNCATE trigger */
+	if (RegularTable(shellRelationId))
+	{
+		CreateTruncateTrigger(shellRelationId);
+	}
+
+	if (ShouldSyncTableMetadata(shellRelationId))
+	{
+		CreateTableMetadataOnWorkers(shellRelationId);
+	}
+
+	/*
+	 * We've a custom way of foreign key graph invalidation,
+	 * see InvalidateForeignKeyGraph().
+	 */
+	if (TableReferenced(shellRelationId) || TableReferencing(shellRelationId))
+	{
+		InvalidateForeignKeyGraph();
+	}
+
+	relation_close(relation, ExclusiveLock);
+}
+
+
+/*
+ * GetShellTableDDLEventsForCitusLocalTable returns a list of DDL commands
+ * to create the shell table from scratch.
+ */
+static List *
+GetShellTableDDLEventsForCitusLocalTable(Oid relationId)
+{
+	List *foreignConstraintCommands =
+		GetForeignConstraintCommandsTableInvolved(relationId);
+
+	/*
+	 * Include DEFAULT clauses for columns getting their default values from
+	 * a sequence.
+	 */
+	bool includeSequenceDefaults = true;
```

Comment:
Hmm, with this one we are slightly diverging from the regular shard creation. On hash distributed tables,  we don't create the sequences on the shards ( see `CreateShardsOnWorkers`), we only have sequences on the coordinator shell tables. And, we evaluate sequences on the coordinator.

As far as I can tell,  this is OK. Let's sign-up @marcocitus's office hour next week, and we explain the latest implementation and can ask these types of edge cases to him.

Reply:
I will look into that Monday

Reply:
Wonder whether we should make a pass over pg_depend to update dependencies to the shell table OID.

Semi-related: https://github.com/citusdata/citus/commit/9ed67a06c8d8dc800e8baa1332632b89dc03918a

Reply:
> onderkalaci 4 days ago Member
Well, it turns out that when there is a sequence in the table, we cannot enable MX:

For the sequences implied by column definitions automatically (serial, bigserial etc), I implemented the following patch:  fbc981ffc05aea8f5e9ce776b8e9a80866daf756 . This commit drops default column definitions from shard relation and grants ownerships of owned sequences to the corresponding columns of the shell table.

However, we will still have two problems around sequences as I mentioned in #3885 & #3884.
They point to existing problems around the other citus table types, and obviously will apply to citus local tables too.

I wonder if I should work on implementing a solution (like the one in 9ed67a06c8d8dc800e8baa1332632b89dc03918a done by Marco) in another pr before merging this pr, or should we leave it to another time ?

(cc: @marcocitus  & cc: @onderkalaci )